src/sci/impl/reflector.cljc

```diff
@@ -0,0 +1,326 @@
+             (identical? param-type Float/TYPE) (.floatValue n)
+             (identical? param-type Double/TYPE) (.doubleValue n)
+             (identical? param-type Long/TYPE) (.longValue n)
+             (identical? param-type Short/TYPE) (.shortValue n)
+             (identical? param-type Byte/TYPE) (.byteValue n)
+             :else (throw (IllegalArgumentException.
+                           (str "Unexpected param type, expected: " param-type
+                                ", given: " (.getName (.getClass arg)))))))
+
+         :else
+         (throw (IllegalArgumentException.
+                 (str "Unexpected param type, expected: " param-type
+                      ", given: " (.getName (.getClass arg)))))))))
+
+#?(:clj
+   (defn- box-args
+     "Box all arguments to match parameter types."
+     [^objects params ^objects args]
+     (if (zero? (alength params))
+       nil
+       (let [ret (object-array (alength params))]
+         (dotimes [i (alength params)]
+           (aset ret i (box-arg (aget params i) (aget args i))))
+         ret))))
+
+#?(:clj
+   (defn- param-arg-type-match?
+     "Check if a parameter type matches an argument type.
+      Includes support for functional interface adaptation."
+     [^Class param-type ^Class arg-type]
+     (cond
+       (nil? arg-type)
+       (not (.isPrimitive param-type))
+
+       (or (= param-type arg-type) (.isAssignableFrom param-type arg-type))
+       true
+
+       (and (maybe-fi-method param-type) (.isAssignableFrom IFn arg-type))
+       true
+
+       (identical? param-type Integer/TYPE)
+       (or (identical? arg-type Integer)
+           (identical? arg-type Long/TYPE)
+           (identical? arg-type Long)
+           (identical? arg-type Short/TYPE)
+           (identical? arg-type Byte/TYPE))
+
+       (identical? param-type Float/TYPE)
+       (or (identical? arg-type Float)
+           (identical? arg-type Double/TYPE))
+
+       (identical? param-type Double/TYPE)
+       (or (identical? arg-type Double)
+           (identical? arg-type Float/TYPE))
+
+       (identical? param-type Long/TYPE)
+       (or (identical? arg-type Long)
+           (identical? arg-type Integer/TYPE)
+           (identical? arg-type Short/TYPE)
+           (identical? arg-type Byte/TYPE))
+
+       (identical? param-type Character/TYPE)
+       (identical? arg-type Character)
+
+       (identical? param-type Short/TYPE)
+       (identical? arg-type Short)
+
+       (identical? param-type Byte/TYPE)
+       (identical? arg-type Byte)
+
+       (identical? param-type Boolean/TYPE)
+       (identical? arg-type Boolean)
+
+       :else
+       false)))
+
+#?(:clj
+   (defn- is-congruent?
+     "Check if parameters are congruent with arguments and arg-types."
+     [^objects params ^objects args ^objects arg-types]
+     (if (nil? args)
+       (zero? (alength params))
+       (and (= (alength params) (alength args))
+            (loop [i 0]
```

Comment:
```suggestion
            ;; TODO refactor to (every? (fn [i] ..) (range (alength params)))
            (loop [i 0]
```

Reply:
Hmm, no refactoring to every? etc will allocate seqs. a loop is more performant and that's important here